app/services/notifications/setup/fetch-download-recipients.service.js

```diff
@@ -0,0 +1,136 @@
+ *
+ * > IMPORTANT! The source for notification contacts is `crm.document_headers` (view `licence_document_headers`), not
+ * > the tables in `crm_v2`.
+ *
+ * Our overall goal is that a 'recipient' receives only one notification, irrespective of how many licences they are
+ * linked to, or what roles they have.
+ *
+ * We start by determining which licences we need to send notifications for, by looking for 'due' return logs with a
+ * matching 'due date' and cycle (summer or winter and all year).
+ *
+ * For each licence linked to one of these return logs, we extract the contact information. This is complicated by a
+ * number of factors.
+ *
+ * - if a licence is _registered_ (more details below), we only care about the email addresses registered against it
+ * - all licences should have a 'licence holder' contact, but they may also have a 'returns' contact
+ * - there is a one-to-one relationship between `licences` and `licence_document_headers`, but the same contact (licence
+ * holder or returns) can appear in different licences, and we are expected to group them into a 'single' contact
+ *
+ * WRLS has the concept of a registered and unregistered licences:
+ *
+ * - **Unregistered licences** have not been linked to an external email, so do not have a 'primary user'. All licences
+ * have a contact with the role 'Licence holder', so this will be extracted as a 'contact'. They may also have a
+ * contact with the role 'Returns to' (but only one), which is extracted as well.
+ * - **Registered licences** have been linked to an external email. That initial email will be linked as the 'primary
+ * user'. These licences may also have designated other accounts as 'returns agents', which will be extracted as well.
+ *
+ * If a licence is registered, we only extract the email contacts. Unregistered licences its the 'Licence holder' and
+ * 'Returns to' contacts from `licence_document_headers.metadata->contacts`.
+ *
+ * We have another service 'FetchContactsService' which removes duplicates rows by squashing them together.
+ * We do not want to remove duplicates for the downloadable recipients. Each row in the CSV file should represent the
+ * data received from this query (For either registered to unregistered licence).
+ * We expect to see duplicate licences with different contacts types
+ * (but still preferring the registered over unregistered licence).
+ *
+ * @param {Date} dueDate
+ * @param {boolean} summer
+ *
+ * @returns {Promise<object[]>} - matching recipients
+ */
+async function go(dueDate, summer) {
+  const { rows } = await _fetch(dueDate, summer)
+
+  return rows
+}
+
+async function _fetch(dueDate, summer) {
+  const query = _query()
+
+  return db.raw(query, [dueDate, summer, dueDate, summer])
+}
+
+function _query() {
+  return `
+SELECT
+  contacts.licence_ref,
+  contacts.contact_type,
+  contacts.return_reference,
```

Comment:
I have mapped the return log data direct. I did not feel it necessary to prefix them with `return_log` but i can see an argument for `return_log_start_date` over just `start_date`. 

happy to change or leave as is.

Reply:
Leave as is